src/nf/nf_random_submodule.f90

```diff
@@ -32,4 +32,22 @@ module function randn_4d(i, j, k, l) result(x)
     x = sqrt(-2 * log(u)) * cos(2 * pi * v)
   end function randn_4d
 
+  module impure elemental subroutine randn_ele(x)
+    real, intent(out) :: x
+    real :: u(2), r
+    real, save :: z
+    logical, save :: have_z = .false.
+    if (have_z) then
+        x = z
+        have_z = .false.
+    else
+        call random_number(u)
+        u(1) = 1.0 - u(1)
+        r = sqrt(-2.0*log(u(1)))
+        x = r*sin(2.0*pi*u(2))
+        z = r*cos(2.0*pi*u(2))
+        have_z = .true.
+    end if
```

Comment:
As I understand it, using saved `z` here effectively saves every call to `random_number` by reusing the sample from the previous call. I think this is generally non-thread safe, and especially that we're running this as `elemental` that could be run in parallel by a smart compiler. Given that computational performance has really low priority (random number sampling is done only once to initialize the weights before the training sequence), I'm in favor of a much simpler code that doesn't keep state, even if slower. That said, what do you think about having just this?

```suggestion
    call random_number(u)
    u(1) = 1.0 - u(1)
    r = sqrt(-2.0*log(u(1)))
    x = r*sin(2.0*pi*u(2))
```